### ISSUE TITLE
Speed-up torch.cat on CPU

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#18634]
## 🚀 Feature
Please speed-up torch.cat on CPU. It should generally be about as fast as a `clone()` call on the output, since both read and write the same number of elements.

It should have good single-threaded and multi-threaded performance.

On large tensors, it seems to be ~4x slower than I would expect.

### Example that does not fit in cache

```python
x = [torch.randn(1, 282240) for _ in range(64)]
%timeit torch.cat(x, dim=0)  # 42 ms -- expected ~9-11 ms

# note that clone is reasonable speed at this size
y = torch.cat(x, dim=0)
%timeit y.clone()  # 11.4 ms
```

**Comment by hanmingliu2** [comment]
@colesbury Hi, I am willing to contribute to this feature. However, after many hours of navigating through the large code base, I am still not sure on what files I should be making changes at. (I am very new to pytorch) May I please have some pointers on that?

**Comment by apaszke** [comment]
@hanmingliu2 When someone calls `torch.cat` it eventually dispatches to [this function](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L173-L181), which in the dense case simply calls into our legacy code which lives [here](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensor.cpp#L690-L696).

**Comment by hanmingliu2** [comment]
@apaszke Thank you so much! Will work on a solution very soon.

**Comment by umanwizard** [comment]
@hanmingliu2 thanks for looking into this, please let us know if you need any help/advice.

**Comment by hanmingliu2** [comment]
@umanwizard Sure. By the way, is it best practice to discuss on how to improve the feature here with you guys before I make an PR?

**Comment by fmassa** [comment]
@hanmingliu2 sure, discussing here before is a good way of starting.

**Comment by hanmingliu2** [comment]
@umanwizard @fmassa After two weeks of examining the source code pointed by @apaszke , I am about to test my hypothesis on this issue. However, I am not able to reproduce the 4x slowdown anymore.

```python
x = [torch.randn(1, 282240) for _ in range(64)]
%timeit torch.cat(x, dim=0)
15.5 ms ± 21.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

y = torch.cat(x, dim=0)
%timeit y.clone()
15.6 ms ± 107 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### On even larger tensors
```python
x = [torch.randn(1, 1000000) for _ in range(64)]
%timeit torch.cat(x, dim=0)
143 ms ± 2.08 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

y = torch.cat(x, dim=0)
%timeit y.clone()
132 ms ± 1.77 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Performance of the two seems to be roughly equivalent. 
Is this issue solved already or there is a mistake in my setup?


### System Info
Pytorch: 1.0.1.post2
Python: 3.7.0
Anaconda: 4.6.11
Intel-MKL: 2019.3
Device: 2017 MacBook Pro
OS: MacOS High Sierra 10.13.6
CPU: Intel Core i5 2.3 GHz
Cores: 2
L2Cache(per core): 256 KB
L3Cache: 4MB 
RAM: 8GB 2133 MHz

**Comment by fmassa** [comment]
@hanmingliu2 can you try updating PyTorch with a nightly version? There has been an update to `clone` that should make it much faster.

**Comment by hanmingliu2** [comment]
@fmassa Thank you so much. The slowdown does appear now after I updated to nightly version.
```python3
import torch
torch.__version__
'1.1.0.dev20190506'

x = [torch.rand(1, 1000000) for _ in range(64)]
%timeit torch.cat(x, dim=0)
139 ms ± 2.22 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

y = torch.cat(x, dim=0)
%timeit y.clone()
78.8 ms ± 2.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by umanwizard** [comment]
@hanmingliu2 are you still working on making `torch.cat` faster?

**Comment by hanmingliu2** [comment]
@umanwizard Yes. I also have an hypothesis on why `torch.cat` is slower than `torch.clone`. However, it's final week of the semester where I have very little time to test such hypothesis. I will have time to work on it  after May 15th, is that okay? I am sorry for the inconvenience my slow progress on this issue may cause.

**Comment by umanwizard** [comment]
@hanmingliu2 Yes, it is definitely okay. There is no inconvenience; we are happy for any help you are able to give :)

**Comment by hanmingliu2** [comment]
@umanwizard I have a few hypotheses which I want to discuss with you guys about:
1. The overheads brought by output size calculation and contiguity check are trivial
2. Slowdown happens when inputs are not all contiguous
3. If **2** is true, then slowdown is in the call `at::_copy_same_type_(nt__wrap, inputs_wrap)` if and only if `at::_copy_same_type_` is slower than `memcpy`(which I think is optimized assembly since C++11?)

What do you guys think? Please correct me on any possible mistake.
Also I am a bit confused on what `THTensor_(narrow)` does. I tried looking for documentation on it but to no avail. May I please have a pointer on that too?

**Comment by fmassa** [comment]
@hanmingliu2 from your previous post, all the inputs that you ded to cat are contiguous, and you are concatenation on the first dimension, so all the slices are also contiguous. Could you clarify what you mean by your point 2 wrt contiguous tensors?

**Comment by hanmingliu2** [comment]
@fmassa Thank you so much for replying. By '**not all contiguous**' I mean that the resulting tensor of calling `torch.cat` may not be contiguous. Such suspicion comes from [this line](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensor.cpp#L773) in the source code.
### My Thought Process
In my understanding, `result` points to the pre-allocated tensor that data is going to be copied into. Then why do we need to check for contiguity on a pre-allocated tensor simply storing resulting data? I then suspect that such need implies that creation of `THTensor` may not always guarantee contiguity. It yet sounds counter-intuitive to myself initially as I believe that PyTorch is using Numpy under the hood which guarantees contiguity on array creation. Though I am not familiar on actual memory allocation mechanism of PyTorch. Therefore, I list point 2 (and thus point 3) in my hypotheses. 

Again please feel free to point out any mistake.

**Comment by umanwizard** [comment]
@hanmingliu2 A few points:

* PyTorch tensors are **not** implemented as Numpy arrays under the hood, though they are conceptually similar.
* It is in fact possible to create non-contiguous tensors in PyTorch, and pass them to `torch.cat`'s `out` parameter:

```python
>>> x=torch.empty(2,3)
>>> y=x[...,0]
>>> y.is_contiguous()
False
>>> a,b=(torch.Tensor([1.0]),torch.Tensor([2.0]))
>>> torch.cat((a,b), out=y)
tensor([1., 2.])
>>> x
tensor([[ 1.0000e+00, -1.9984e+18, -1.9984e+18],
        [ 2.0000e+00, -1.9984e+18, -1.9984e+18]])
```

**Comment by hanmingliu2** [comment]
@umanwizard Thank you so much for clarifying. I am currently reading through the related parts of the code base to understand the mechanism of `torch.cat` better.

**Comment by hanmingliu2** [comment]
@umanwizard Due to heavy work schedule, I work on this issue only during my free time. Again I am sorry for my slow progress. I will notify you guys as soon as I make new progress on resolving this issue.
